File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/F-Закраска_прямой/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    """
    n: int = int(sys.stdin.readline().rstrip())
    segments: list[tuple[int, int]] = []
    for _ in range(n):
        start, end = map(int, sys.stdin.readline().rstrip().split())
        segments.append((start, end))
    segments.sort()
    output: int = 0
    prev_start: int = -2 * 10**9
    prev_end: int = -2 * 10**9
    for start, end in segments:
        if start <= prev_end: prev_end = max(prev_end, end)
        else:
            output += prev_end - prev_start
            prev_start, prev_end = start, end
    output += prev_end - prev_start
    sys.stdout.write('{}\n'.format(output))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/4sprint-SolvingProblemMethodsPart1/contest/F-Закраска_прямой/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdint>


void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(1)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<std::pair<int, int>> segments;
    for (int i = 0; i < n; ++i) {
        int start, end;
        std::scanf("%d %d", &start, &end);
        segments.push_back({start, end});
    }
    std::sort(segments.begin(), segments.end());
    long long output = 0;
    int prevStart = INT32_MIN, prevEnd = INT32_MIN;
    for (std::pair<int, int>& segment : segments) {
        if (segment.first <= prevEnd) prevEnd = std::max(prevEnd, segment.second);
        else {
            output += (long long)prevEnd - (long long)prevStart;
            prevStart = segment.first;
            prevEnd = segment.second;
        }
    }
    output += (long long)prevEnd - (long long)prevStart;
    std::printf("%lld\n", output);
}


int main() {
    solution();
}